keras_cv/metrics/coco/pycoco_utils.py

```diff
@@ -0,0 +1,231 @@
+        batch_size = groundtruths["source_id"][i].shape[0]
+        for j in range(batch_size):
+            num_instances = groundtruths["num_detections"][i][j]
+            if num_instances > max_num_instances:
+                num_instances = max_num_instances
+            for k in range(int(num_instances)):
+                ann = {}
+                ann["image_id"] = groundtruths["source_id"][i][j]
+                ann["iscrowd"] = 0
+                ann["category_id"] = int(groundtruths["classes"][i][j, k])
+                boxes = groundtruths["boxes"][i]
+                ann["bbox"] = [
+                    float(boxes[j, k, 1]),
+                    float(boxes[j, k, 0]),
+                    float(boxes[j, k, 3] - boxes[j, k, 1]),
+                    float(boxes[j, k, 2] - boxes[j, k, 0]),
+                ]
+                ann["area"] = float(
+                    (boxes[j, k, 3] - boxes[j, k, 1])
+                    * (boxes[j, k, 2] - boxes[j, k, 0])
+                )
+                gt_annotations.append(ann)
+
+    for i, ann in enumerate(gt_annotations):
+        ann["id"] = i + 1
+
+    if label_map:
+        gt_categories = [{"id": i, "name": label_map[i]} for i in label_map]
+    else:
+        category_ids = [gt["category_id"] for gt in gt_annotations]
+        gt_categories = [{"id": i} for i in set(category_ids)]
+
+    gt_dataset = {
+        "images": gt_images,
+        "categories": gt_categories,
+        "annotations": copy.deepcopy(gt_annotations),
+    }
+    return gt_dataset
+
+
+def _convert_to_numpy(groundtruths, predictions):
+    """Converts tensors to numpy arrays."""
+    labels = tf.nest.map_structure(lambda x: x.numpy(), groundtruths)
+    numpy_groundtruths = {}
+    for key, val in labels.items():
+        if isinstance(val, tuple):
+            val = np.concatenate(val)
+        numpy_groundtruths[key] = val
+
+    outputs = tf.nest.map_structure(lambda x: x.numpy(), predictions)
+    numpy_predictions = {}
+    for key, val in outputs.items():
+        if isinstance(val, tuple):
+            val = np.concatenate(val)
+        numpy_predictions[key] = val
+
+    return numpy_groundtruths, numpy_predictions
+
+
+def compute_pycoco_metrics(groundtruths, predictions):
+    groundtruths, predictions = _convert_to_numpy(groundtruths, predictions)
+
+    gt_dataset = _convert_groundtruths_to_coco_dataset(groundtruths)
+    coco_gt = COCOWrapper(gt_dataset=gt_dataset)
+    coco_predictions = _convert_predictions_to_coco_annotations(predictions)
+    coco_dt = coco_gt.loadRes(predictions=coco_predictions)
+    image_ids = [ann["image_id"] for ann in coco_predictions]
+
+    coco_eval = cocoeval.COCOeval(coco_gt, coco_dt, iouType="bbox")
+    coco_eval.params.imgIds = image_ids
+    coco_eval.evaluate()
+    coco_eval.accumulate()
+    coco_eval.summarize()
+    coco_metrics = coco_eval.stats
+
+    metrics = coco_metrics
+
+    metrics_dict = {}
+    for i, name in enumerate(METRIC_NAMES):
```

Comment:
do we need this given coco_eval.summarize() already gives the expected log?

Reply:
Just verified this, and `coco_metrics` here is a list of numbers (without any associated metric names) -- so this is required to associated the metric names with the metrics for the output logs

Reply:
last time I tried it, it gives something like AP50, AP75, APSmall, APLarge, etc, is that not the case anymore?

Reply:
coco_eval.eval gives a dictionary with e.g. `recall` and `precision`, but on a per-image basis. stats just gives an array. Implementation [here](https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocotools/cocoeval.py).